### PR TITLE
Property serializer

### PR DESCRIPTION
**Pull request by apopiak**
serialize property values to JSON + populate properties from JSON

### CODE REVIEW

File: source/reflectionzeug/source/JsonPropertyDeserializer.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+{
+    if (m_groupStack.back()->name().compare(m_currentGroupName) == 0)
+    {
+        m_groupStack.pop_back();
+        if (!m_groupStack.empty())
+        {
+            m_currentGroupName = m_groupStack.back()->name();
+        }
+    }
+}
```
Comment (cgcostume) [inline source/reflectionzeug/source/JsonPropertyDeserializer.cpp]: reduce nesting level:
```
if (m_groupStack.back()->name().compare(m_currentGroupName) != 0)
   return;

m_groupStack.pop_back();
if (!m_groupStack.empty())
    m_currentGroupName = m_groupStack.back()->name();
```
... and is it assured that there is a back?

File: source/reflectionzeug/source/JsonPropertySerializer.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            m_nestingLevel--;
+            m_fstream << indent(m_nestingLevel) << "}";
+            if (m_elementCount.at(m_nestingLevel) > 0)
+            {
+                m_fstream << ",";
+            }
+            m_fstream << std::endl;
+            m_elementCount.pop_back();
+        };
+}
```
Comment (cgcostume) [inline source/reflectionzeug/source/JsonPropertySerializer.cpp]: i do not understand why it is necessary to use lambda expressions here - why not just use functions?
Especially: this is a recursive use of lambdas? again, why not use plain old functions ...

File: source/reflectionzeug/source/JsonPropertyDeserializer.cpp [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+    for (std::string line; std::getline(fstream, line);)
+    {
+        if (this->isGroupDeclaration(line))
+            noErrorsOccured = this->updateCurrentGroup(line) && noErrorsOccured;
+        else if (this->isGroupEnd(line))
+            this->endGroup();
+        else if (this->isPropertyDeclaration(line))
+            noErrorsOccured = this->setPropertyValue(line) && noErrorsOccured;
+    }
```
Comment (cgcostume) [inline source/reflectionzeug/source/JsonPropertyDeserializer.cpp]: it should not be required to use ```this->``` here

File: source/reflectionzeug/source/JsonPropertyDeserializer.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+#include <reflectionzeug/util.h>
+
+using namespace loggingzeug;
+
+namespace reflectionzeug
+{
+
+JsonPropertyDeserializer::JsonPropertyDeserializer()
+:   m_rootGroup(nullptr)
+{
```
Comment (cgcostume) [inline source/reflectionzeug/source/JsonPropertyDeserializer.cpp]: prefer uniform initialization: ```m_rootGroup{ nullptr }```

File: source/reflectionzeug/include/reflectionzeug/VariantConverterInit.h [hunk 5]
```diff
@@ -112,6 +113,21 @@ struct VariantConverterInit<ValueType, EnableIf<std::is_floating_point<ValueType
...
+{
+    void operator()()
+    {
+        Variant::registerConverter<FilePath, std::string>(
+            [](const FilePath & path) -> std::string
+                {
+                    std::string str(path.toString());
+                    return str;
+                }
+        );
```
Comment (cgcostume) [inline source/reflectionzeug/include/reflectionzeug/VariantConverterInit.h]: indentation
```
Variant::registerConverter<FilePath, std::string>(
    [](const FilePath & path) -> std::string
    {
        std::string str(path.toString());
        return str;
    }
```

File: source/reflectionzeug/source/IniPropertyDeserializer.cpp [hunk 6]
```diff
@@ -24,23 +24,23 @@ using namespace loggingzeug;
 namespace reflectionzeug
 {
 
-PropertyDeserializer::PropertyDeserializer()
+IniPropertyDeserializer::IniPropertyDeserializer()
 :   m_rootGroup(nullptr)
 ,   m_currentGroup(nullptr)
 ,   m_currentValue("")
 {
```
Comment (cgcostume) [inline source/reflectionzeug/source/IniPropertyDeserializer.cpp]: prefer uniform initialization

File: source/reflectionzeug/source/IniPropertyDeserializer.cpp [hunk 7]
```diff
@@ -110,17 +110,17 @@ bool PropertyDeserializer::setPropertyValue(const std::string & line)
...
 
     if (!property) {
-        critical() << "Property path \"" << path << "\" " << "is invalid" << std::endl;
+        warning() << "Property path \"" << path << "\" " << "is invalid" << std::endl;
         return false;
     }
 
     if (!property->isValue()) {
-        critical() << "Tried to assign value to group with name: " << property->name() << std::endl;
+        warning() << "Tried to assign value to group with name: " << property->name() << std::endl;
```
Comment (cgcostume) [inline source/reflectionzeug/source/IniPropertyDeserializer.cpp]: prefer extra line for beginning of a block
```
if ()
{
...
}
```

File: source/reflectionzeug/source/IniPropertySerializer.cpp [hunk 8]
```diff
@@ -75,17 +75,17 @@ void PropertySerializer::serializeGroup(const PropertyGroup & group)
         });
```
Comment (cgcostume) [inline source/reflectionzeug/source/IniPropertySerializer.cpp]: i'm unsure about the overall design and use of lambdas - can you please provide details for the design decision

File: source/reflectionzeug/source/JsonPropertyDeserializer.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    std::fstream fstream;
+    fstream.open(filePath, std::ios_base::in);
+    if (!fstream.is_open())
+    {
+        warning() << "Could not open file \"" << filePath << "\"" << std::endl;
+        return false;
+    }
+
+    bool noErrorsOccured = true;
```
Comment (cgcostume) [inline source/reflectionzeug/source/JsonPropertyDeserializer.cpp]: prefer ```auto```

File: source/reflectionzeug/source/IniPropertyDeserializer.cpp [hunk 10]
```diff
@@ -58,14 +58,14 @@ bool PropertyDeserializer::deserialize(PropertyGroup & group, const std::string
     return noErrorsOccured;
 }
 
-bool PropertyDeserializer::isGroupDeclaration(const std::string & line)
+bool IniPropertyDeserializer::isGroupDeclaration(const std::string & line)
 {
     static const regex_namespace::regex groupRegex("\\[" + AbstractProperty::s_nameRegexString + "\\]");
```
Comment (cgcostume) [inline source/reflectionzeug/source/IniPropertyDeserializer.cpp]: prefer auto syntax

Comment (scheibel) [inline source/reflectionzeug/source/IniPropertyDeserializer.cpp]: seems like ```auto``` isn't applicable here, but uniform initialization is.

**Comment by apopiak** [comment]
Thanks for all the comments, but the PR is closed because we are refactoring a big part of it. The Lambda was useful (if not necessary) because it made the `group->forEachGroup()` call easier (we would have had to add a context to a function, so the lambda was an easier way to achieve this).
The recursion was the (I think) cleanest and easiest solution.

File: source/reflectionzeug/source/IniPropertyDeserializer.cpp (hunk 10 above)
Comment (mjendruk) [inline source/reflectionzeug/source/IniPropertyDeserializer.cpp]: I think both are applicable.
```c++
static const auto groupRegex = regex_namespace::regex{"\\[" + AbstractProperty::s_nameRegexString + "\\]"};
```

**Pull request closed without merging**